File: 07_TA1092_Pembro_ endometrial cancer/BkupKMDigi/dMMR_PFS_Bkup/guyot_style_reconstruct.py
#!/usr/bin/env python3
import csv
import json
import math
from pathlib import Path


HERE = Path(__file__).resolve().parent


def read_segments(path):
    rows = []
    with open(path, newline="") as f:
        reader = csv.DictReader(f)
        for row in reader:
            rows.append(
                {
                    "arm": row["arm"],
                    "start": float(row["start_time_months"]),
                    "end": float(row["end_time_months"]),
                    "surv": float(row["survival"]),
                }
            )
    rows.sort(key=lambda x: (x["start"], x["end"]))
    return rows


def event_drops_from_segments(segments, min_drop=0.002):
    drops = []
    previous_surv = 1.0
    for seg in segments:
        drop = previous_surv - seg["surv"]
        if drop >= min_drop:
            drops.append(
                {
                    "time": max(0.0, seg["start"]),
                    "surv_before": previous_surv,
                    "surv_after": seg["surv"],
                    "drop": drop,
                }
            )
        previous_surv = min(previous_surv, seg["surv"])
    return drops


def interval_index(t, interval_times):
    for i in range(len(interval_times) - 1):
        if interval_times[i] <= t < interval_times[i + 1]:
            return i
    return len(interval_times) - 2


def reconstruct_arm(arm_name, segment_path, nar_times, nar_values, target_events):
    segments = read_segments(segment_path)
    drops = event_drops_from_segments(segments)

    # Estimate event weights interval by interval, then constrain total event count to
    # the published arm-level number of events. This avoids treating anti-aliased
    # micro-drops as extra events.
    weighted_drops = []
    running_risk = {i: nar_values[i] for i in range(len(nar_times) - 1)}
    for drop in drops:
        i = interval_index(drop["time"], nar_times)
        ratio = max(0.0, min(1.0, drop["surv_after"] / drop["surv_before"]))
        raw_events = max(0.0, running_risk[i] * (1.0 - ratio))
        if raw_events > 0:
            weighted_drops.append({**drop, "interval": i, "raw_events": raw_events})
            running_risk[i] = max(0.0, running_risk[i] - raw_events)

    for d in weighted_drops:
        d["events"] = int(round(d["raw_events"]))

    raw_event_total = sum(d["events"] for d in weighted_drops)
    excess = raw_event_total - target_events
    if excess > 0:
        # Preserve clinically important KM drops and remove excess events
        # preferentially from the smallest visual survival drops that are more
        # likely to reflect anti-aliasing or censor-mark contamination. A late
        # event may have low raw event weight because few patients remain at
        # risk, but it can still create a large KM step and should be retained.
        ranked = sorted(
            range(len(weighted_drops)),
            key=lambda k: (weighted_drops[k]["drop"], -weighted_drops[k]["time"]),
        )
        while excess > 0:
            changed = False
            for k in ranked:
                if excess <= 0:
                    break
                if weighted_drops[k]["events"] > 0:
                    weighted_drops[k]["events"] -= 1
                    excess -= 1
                    changed = True
            if not changed:
                break
    elif excess < 0:
        deficit = -excess
        # Add missing events to the largest observed drops first.
        ranked = sorted(
            range(len(weighted_drops)),
            key=lambda k: weighted_drops[k]["raw_events"],
            reverse=True,
        )
        i = 0
        while deficit > 0 and ranked:
            weighted_drops[ranked[i % len(ranked)]]["events"] += 1
            deficit -= 1
            i += 1

    # Allocate event counts interval by interval. Censoring is placed after observed
    # event drops within each NAR interval to force agreement with published risk sets.
    ipd = []
    event_rows = []
    censor_rows = []
    next_id = 1
    total_events = 0

    drops_by_interval = {i: [] for i in range(len(nar_times) - 1)}
    for drop in weighted_drops:
        drops_by_interval[drop["interval"]].append(drop)

    for i in range(len(nar_times) - 1):
        start = nar_times[i]
        end = nar_times[i + 1]
        nrisk = nar_values[i]
        desired_next = nar_values[i + 1]
        interval_events = []

        for drop in drops_by_interval[i]:
            d = int(drop["events"])
            d = max(0, min(d, nrisk))
            if d > 0:
                interval_events.append((drop["time"], d))
                nrisk -= d
                total_events += d

        # Force the next risk-set count. This is the key NAR constraint.
        c = max(0, nrisk - desired_next)
        if c > 0:
            # Spread censoring after the last event and before the next NAR point.
            last_event_t = max([t for t, _ in interval_events], default=start)
            c_start = min(end, max(start, last_event_t + 1e-4))
            if c == 1:
                censor_times = [(c_start + end) / 2.0]
            else:
                censor_times = [
                    c_start + (end - c_start) * (k + 1) / (c + 1)
                    for k in range(c)
                ]
        else:
            censor_times = []

        for t, d in interval_events:
            for _ in range(d):
                ipd.append(
                    {
                        "id": f"{arm_name}_{next_id}",
                        "arm": arm_name,
                        "time": t,
                        "status": 1,
                    }
                )
                next_id += 1
            event_rows.append({"interval": i, "time": t, "events": d})

        for t in censor_times:
            ipd.append(
                {
                    "id": f"{arm_name}_{next_id}",
                    "arm": arm_name,
                    "time": t,
                    "status": 0,
                }
            )
            next_id += 1
        censor_rows.append({"interval": i, "censors": len(censor_times)})

    # Censor anyone still at risk after the final NAR time at the last observed curve time.
    remaining = nar_values[-1]
    last_time = max([s["end"] for s in segments] + [nar_times[-1]])
    for _ in range(remaining):
        ipd.append(
            {
                "id": f"{arm_name}_{next_id}",
                "arm": arm_name,
                "time": last_time,
                "status": 0,
            }
        )
        next_id += 1

    # Total N can drift if an interval has impossible event/censor allocation after rounding.
    # Keep this visible rather than silently correcting it.
    validation = validate_arm(ipd, nar_times, nar_values, target_events)
    validation["estimated_events_before_event_count_adjustment"] = total_events

    return ipd, event_rows, censor_rows, validation


def validate_arm(ipd, nar_times, nar_values, target_events):
    validation_rows = []
    for t, published in zip(nar_times, nar_values):
        # KM convention: at risk just before t includes patients with event/censor at t.
        reconstructed = sum(1 for row in ipd if row["time"] >= t - 1e-9)
        validation_rows.append(
            {
                "time_months": t,
                "published_nar": published,
                "reconstructed_nar": reconstructed,
                "difference": reconstructed - published,
            }
        )
    events = sum(row["status"] for row in ipd)
    return {
        "n": len(ipd),
        "events": events,
        "target_events": target_events,
        "event_difference": events - target_events,
        "nar_validation": validation_rows,
        "median_months": km_median(ipd),
    }


def km_median(ipd):
    n = len(ipd)
    if n == 0:
        return None
    at_risk = n
    surv = 1.0
    all_times = sorted({row["time"] for row in ipd})
    for t in all_times:
        d = sum(1 for row in ipd if row["time"] == t and row["status"] == 1)
        c = sum(1 for row in ipd if row["time"] == t and row["status"] == 0)
        if d > 0 and at_risk > 0:
            surv *= 1.0 - d / at_risk
        if d > 0 and surv <= 0.5:
            return t
        at_risk -= d + c
    return None


def write_csv(path, rows, columns):
    with open(path, "w", newline="") as f:
        writer = csv.DictWriter(f, fieldnames=columns)
        writer.writeheader()
        writer.writerows(rows)


def main():
    nar = json.loads((HERE / "dMMR_PFS_nar.json").read_text())
    nar_times = nar["time_months"]

    segment_paths = {
        "pembro_chemo": HERE / "dMMR_PFS_pembro_step_segments.csv",
        "placebo_chemo": HERE / "dMMR_PFS_placebo_step_segments_revised.csv",
    }

    all_ipd = []
    report = {}
    for arm, path in segment_paths.items():
        meta = nar["arms"][arm]
        ipd, event_rows, censor_rows, validation = reconstruct_arm(
            arm,
            path,
            nar_times,
            meta["nar"],
            meta["events_reported"],
        )
        all_ipd.extend(ipd)
        write_csv(HERE / f"{arm}_dMMR_PFS_pseudo_ipd.csv", ipd, ["id", "arm", "time", "status"])
        write_csv(HERE / f"{arm}_event_allocation.csv", event_rows, ["interval", "time", "events"])
        write_csv(HERE / f"{arm}_censor_allocation.csv", censor_rows, ["interval", "censors"])
        report[arm] = validation

    write_csv(HERE / "dMMR_PFS_pseudo_ipd_combined.csv", all_ipd, ["id", "arm", "time", "status"])
    (HERE / "dMMR_PFS_guyot_style_validation.json").write_text(
        json.dumps(report, indent=2), encoding="utf-8"
    )
    print(json.dumps(report, indent=2))


if __name__ == "__main__":
    main()

File: 07_TA1092_Pembro_ endometrial cancer/BkupKMDigi/dMMR_PFS_Bkup/guyot_style_reconstruct.R
library(survival)
library(jsonlite)

`%||%` <- function(x, y) if (is.null(x) || length(x) == 0) y else x

read_segments <- function(path) {
  x <- read.csv(path, stringsAsFactors = FALSE)
  names(x) <- c("arm", "start", "end", "surv")
  x[order(x$start, x$end), ]
}

read_nar_json <- function(path) {
  fromJSON(path, simplifyVector = TRUE)
}

event_drops_from_segments <- function(segments, min_drop = 0.002) {
  previous <- 1.0
  out <- list()
  for (i in seq_len(nrow(segments))) {
    drop <- previous - segments$surv[i]
    if (drop >= min_drop) {
      out[[length(out) + 1]] <- data.frame(
        time = max(0, segments$start[i]),
        surv_before = previous,
        surv_after = segments$surv[i],
        drop = drop
      )
    }
    previous <- min(previous, segments$surv[i])
  }
  if (length(out) == 0) {
    data.frame(time = numeric(), surv_before = numeric(), surv_after = numeric(), drop = numeric())
  } else {
    do.call(rbind, out)
  }
}

interval_index <- function(t, interval_times) {
  idx <- findInterval(t, interval_times, rightmost.closed = FALSE)
  pmin(pmax(idx, 1), length(interval_times) - 1)
}

km_median_manual <- function(ipd) {
  at_risk <- nrow(ipd)
  surv <- 1.0
  for (tt in sort(unique(ipd$time))) {
    d <- sum(ipd$time == tt & ipd$status == 1)
    c <- sum(ipd$time == tt & ipd$status == 0)
    if (d > 0 && at_risk > 0) {
      surv <- surv * (1 - d / at_risk)
      if (surv <= 0.5) return(tt)
    }
    at_risk <- at_risk - d - c
  }
  NA_real_
}

reconstruct_arm <- function(arm_name, segment_path, nar_times, nar_values, target_events) {
  segments <- read_segments(segment_path)
  drops <- event_drops_from_segments(segments)

  weighted <- list()
  running_risk <- nar_values[-length(nar_values)]
  if (nrow(drops) > 0) {
    for (i in seq_len(nrow(drops))) {
      int_idx <- interval_index(drops$time[i], nar_times)
      ratio <- max(0, min(1, drops$surv_after[i] / drops$surv_before[i]))
      raw_events <- max(0, running_risk[int_idx] * (1 - ratio))
      if (raw_events > 0) {
        weighted[[length(weighted) + 1]] <- data.frame(
          time = drops$time[i],
          surv_before = drops$surv_before[i],
          surv_after = drops$surv_after[i],
          drop = drops$drop[i],
          interval = int_idx,
          raw_events = raw_events,
          events = round(raw_events)
        )
        running_risk[int_idx] <- max(0, running_risk[int_idx] - raw_events)
      }
    }
  }
  weighted <- if (length(weighted) == 0) {
    data.frame(time = numeric(), surv_before = numeric(), surv_after = numeric(), drop = numeric(),
               interval = integer(), raw_events = numeric(), events = integer())
  } else {
    do.call(rbind, weighted)
  }

  excess <- sum(weighted$events) - target_events
  if (nrow(weighted) > 0 && excess > 0) {
    ranked <- order(weighted$drop, -weighted$time)
    while (excess > 0 && any(weighted$events > 0)) {
      for (k in ranked) {
        if (excess <= 0) break
        if (weighted$events[k] > 0) {
          weighted$events[k] <- weighted$events[k] - 1
          excess <- excess - 1
        }
      }
    }
  } else if (nrow(weighted) > 0 && excess < 0) {
    deficit <- -excess
    ranked <- order(weighted$raw_events, decreasing = TRUE)
    i <- 1
    while (deficit > 0) {
      k <- ranked[((i - 1) %% length(ranked)) + 1]
      weighted$events[k] <- weighted$events[k] + 1
      deficit <- deficit - 1
      i <- i + 1
    }
  }

  ipd <- data.frame(id = character(), arm = character(), time = numeric(), status = integer())
  next_id <- 1
  event_rows <- data.frame(interval = integer(), time = numeric(), events = integer())
  censor_rows <- data.frame(interval = integer(), censors = integer())

  for (i in seq_len(length(nar_times) - 1)) {
    start <- nar_times[i]
    end <- nar_times[i + 1]
    nrisk <- nar_values[i]
    desired_next <- nar_values[i + 1]
    w <- weighted[weighted$interval == i, , drop = FALSE]
    interval_events <- data.frame(time = numeric(), events = integer())

    if (nrow(w) > 0) {
      for (j in seq_len(nrow(w))) {
        d <- max(0, min(w$events[j], nrisk))
        if (d > 0) {
          interval_events <- rbind(interval_events, data.frame(time = w$time[j], events = d))
          nrisk <- nrisk - d
        }
      }
    }

    c_count <- max(0, nrisk - desired_next)
    last_event_t <- if (nrow(interval_events) > 0) max(interval_events$time) else start
    c_start <- min(end, max(start, last_event_t + 1e-4))
    censor_times <- if (c_count == 0) {
      numeric()
    } else if (c_count == 1) {
      (c_start + end) / 2
    } else {
      c_start + (end - c_start) * seq_len(c_count) / (c_count + 1)
    }

    if (nrow(interval_events) > 0) {
      for (j in seq_len(nrow(interval_events))) {
        for (k in seq_len(interval_events$events[j])) {
          ipd <- rbind(ipd, data.frame(
            id = sprintf("%s_%d", arm_name, next_id),
            arm = arm_name,
            time = interval_events$time[j],
            status = 1
          ))
          next_id <- next_id + 1
        }
      }
      event_rows <- rbind(event_rows, transform(interval_events, interval = i))
    }

    for (tt in censor_times) {
      ipd <- rbind(ipd, data.frame(
        id = sprintf("%s_%d", arm_name, next_id),
        arm = arm_name,
        time = tt,
        status = 0
      ))
      next_id <- next_id + 1
    }
    censor_rows <- rbind(censor_rows, data.frame(interval = i, censors = length(censor_times)))
  }

  remaining <- nar_values[length(nar_values)]
  last_time <- max(c(segments$end, nar_times[length(nar_times)]))
  if (remaining > 0) {
    for (k in seq_len(remaining)) {
      ipd <- rbind(ipd, data.frame(
        id = sprintf("%s_%d", arm_name, next_id),
        arm = arm_name,
        time = last_time,
        status = 0
      ))
      next_id <- next_id + 1
    }
  }

  validation <- data.frame(
    time_months = nar_times,
    published_nar = nar_values,
    reconstructed_nar = sapply(nar_times, function(tt) sum(ipd$time >= tt - 1e-9))
  )
  validation$difference <- validation$reconstructed_nar - validation$published_nar

  list(
    ipd = ipd,
    event_rows = event_rows,
    censor_rows = censor_rows,
    validation = validation,
    summary = data.frame(
      arm = arm_name,
      n = nrow(ipd),
      events = sum(ipd$status),
      target_events = target_events,
      event_difference = sum(ipd$status) - target_events,
      median_months = km_median_manual(ipd)
    )
  )
}

run_reconstruction <- function(base_dir = getwd()) {
  nar <- read_nar_json(file.path(base_dir, "dMMR_PFS_nar.json"))
  specs <- list(
    pembro_chemo = list(path = "dMMR_PFS_pembro_step_segments.csv", meta = nar$arms$pembro_chemo),
    placebo_chemo = list(path = "dMMR_PFS_placebo_step_segments_revised.csv", meta = nar$arms$placebo_chemo)
  )

  results <- list()
  all_ipd <- data.frame(id = character(), arm = character(), time = numeric(), status = integer())
  for (arm in names(specs)) {
    res <- reconstruct_arm(
      arm,
      file.path(base_dir, specs[[arm]]$path),
      nar$time_months,
      specs[[arm]]$meta$nar,
      specs[[arm]]$meta$events_reported
    )
    results[[arm]] <- res
    all_ipd <- rbind(all_ipd, res$ipd)
    write.csv(res$ipd, file.path(base_dir, sprintf("%s_dMMR_PFS_pseudo_ipd_R.csv", arm)), row.names = FALSE)
    write.csv(res$validation, file.path(base_dir, sprintf("%s_nar_validation_R.csv", arm)), row.names = FALSE)
  }
  write.csv(all_ipd, file.path(base_dir, "dMMR_PFS_pseudo_ipd_combined_R.csv"), row.names = FALSE)

  all_ipd$arm <- factor(all_ipd$arm, levels = c("placebo_chemo", "pembro_chemo"))
  fit <- survfit(Surv(time, status) ~ arm, data = all_ipd)
  cox <- coxph(Surv(time, status) ~ arm, data = all_ipd)
  hr <- exp(coef(cox))[1]
  ci <- exp(confint(cox))[1, ]
  med <- summary(fit)$table[, "median"]
  stats <- data.frame(
    metric = c("Cox HR pembro vs placebo", "HR CI lower", "HR CI upper", "Median placebo", "Median pembro"),
    value = c(hr, ci[1], ci[2], med["arm=placebo_chemo"], med["arm=pembro_chemo"]),
    target = c(0.30, 0.19, 0.48, 7.6, NA)
  )
  write.csv(stats, file.path(base_dir, "dMMR_PFS_statistical_validation_R.csv"), row.names = FALSE)
  print(do.call(rbind, lapply(results, `[[`, "summary")))
  print(stats)
}

args <- commandArgs(trailingOnly = TRUE)
run_reconstruction(if (length(args) >= 1) args[1] else getwd())
